Add config file for openshift-ansible install

spadgett opened the pull request:
Needed for https://github.com/openshift/openshift-ansible/pull/8328

Adds gopkg.in/yaml.v2 dependency for YAML parsing (as a separate commit to make review easier)

spadgett commented:
@ericchiang fyi

Do you know if I re-vendored correctly? The lines changed is much larger than I'd expect :(

spadgett commented:
jenkins push

coreos-ui commented:
The PR image has been pushed to Quay with tag `pr_65_build_293`

ericchiang commented:
@spadgett you probably need to install https://github.com/sgotti/glide-vc

We should probably just switch to dep after this :)

ericchiang commented:
Also I would _strongly_ recommend using https://github.com/ghodss/yaml in combination with [`encoding/json.Decoder.DisallowUnknownFields`](https://golang.org/pkg/encoding/json/#Decoder.DisallowUnknownFields) to make it an error to provide an unknown field. YAML parsing dropping incorrectly indented fields has been a huge issue for us in https://github.com/coreos/dex and I wish we had done that from the beginning.

spadgett commented:
jenkins push

coreos-ui commented:
The PR image failed.

spadgett commented:
jenkins push

coreos-ui commented:
The PR image failed.

spadgett commented:
jenkins push

coreos-ui commented:
The PR image failed.

spadgett commented:
jenkins push

coreos-ui commented:
The PR image failed.

ggreer commented:
jenkins push

ggreer commented on cmd/bridge/config.go:
```diff
@@ -0,0 +1,120 @@
+
+// Customization holds configuration such as what logo to use.
+type Customization struct {
+	LogoImageName string `yaml:"logoImageName"`
```
IMO we should drop this "feature". It's nothing but trouble.

coreos-ui commented:
The PR image has been pushed to Quay with tag `pr_65_build_299`

stevekuznetsov commented:
/test backend
/test frontend

spadgett commented:
jenkins push

coreos-ui commented:
The PR image has been pushed to Quay with tag `pr_65_build_386`

spadgett commented:
/test backend

spadgett commented:
@liggitt @deads2k are one of you able to give this an API review?

@jwforres fyi

spadgett commented:
Example config:

```yaml
apiVersion: console.openshift.io/v1beta1
kind: ConsoleConfig
servingInfo:
  bindAddress: https://0.0.0.0:8443 
  certFile: /var/serving-cert/tls.crt
  keyFile: /var/serving-cert/tls.key
auth:
  clientID: openshift-console
  clientSecretFile: /var/oauth-config/clientSecret
  issuerURL: https://public-master.example.com:8443 
clusterInfo:
  consoleBaseAddress: https://console.router.default.svc.cluster.local
customization:
  documentationBaseURL: https://docs.openshift.org/3.11/ 
  logoImageName: origin
```

spadgett commented:
jenkins push

coreos-ui commented:
The PR image has been pushed to Quay with tag `pr_65_build_390`

spadgett commented:
jenkins push

spadgett commented:
@deads2k I added a separate `issuerCAFile`

coreos-ui commented:
The PR image has been pushed to Quay with tag `pr_65_build_397`

enj commented on cmd/bridge/config.go:
```diff
@@ -0,0 +1,184 @@
+
+// Auth holds configuration for authenticating with OpenShift. The auth method is assumed to be "openshift".
+type Auth struct {
+	IssuerURL        string `yaml:"issuerURL"`
```
What is this used for?

spadgett replied:
```
wellKnownURL := strings.TrimSuffix(c.issuerURL, "/") + "/.well-known/oauth-authorization-server"
```

https://github.com/openshift/console/blob/612f933e2cab16af2fdeecdb5cd33df0a62eb9b0/auth/auth_openshift.go#L31

spadgett replied:
OK, I plan to remove this and use the kubernetes service in-cluster and master public URL off-cluster for getting the well-known doc

enj commented on cmd/bridge/config.go:
```diff
@@ -0,0 +1,184 @@
+// Auth holds configuration for authenticating with OpenShift. The auth method is assumed to be "openshift".
+type Auth struct {
+	IssuerURL        string `yaml:"issuerURL"`
+	IssuerCAFile     string `yaml:"issuerCAFile"`
```
I assume we still have no solution for the routing problem on CI?

I assume we do *not* want to allow custom override of the OAuth metadata similar to the master:

```go
	// OAuthMetadataFile is a path to a file containing the discovery endpoint for OAuth 2.0 Authorization
	// Server Metadata for an external OAuth server.
	// See IETF Draft: // https://tools.ietf.org/html/draft-ietf-oauth-discovery-04#section-2
	// This option is mutually exclusive with OAuthConfig
	OAuthMetadataFile string `json:"oauthMetadataFile"`
```

enj commented on cmd/bridge/config.go:
```diff
@@ -0,0 +1,184 @@
+	IssuerURL        string `yaml:"issuerURL"`
+	IssuerCAFile     string `yaml:"issuerCAFile"`
+	ClientID         string `yaml:"clientID"`
+	ClientSecretFile string `yaml:"clientSecretFile"`
```
Does this need to support encryption on disk?

enj commented on cmd/bridge/config.go:
```diff
@@ -0,0 +1,184 @@
+	IssuerCAFile     string `yaml:"issuerCAFile"`
+	ClientID         string `yaml:"clientID"`
+	ClientSecretFile string `yaml:"clientSecretFile"`
+	LogoutRedirect   string `yaml:"logoutRedirect"`
```
Where is supposed to go "normally"?

spadgett replied:
Back to the console if no redirect set.

deads2k commented on cmd/bridge/config.go:
```diff
@@ -0,0 +1,184 @@
+}
+
+// Auth holds configuration for authenticating with OpenShift. The auth method is assumed to be "openshift".
+type Auth struct {
```
I thought you were adding a CA bundle file to this.

spadgett commented:
jenkins push

coreos-ui commented:
The PR image failed.

spadgett commented:
jenkins push

coreos-ui commented:
The PR image failed.

spadgett commented:
jenkins push

spadgett commented:
OK, this is updated:

* Removed `issuerURL` from `auth` stanza (not needed for OpenShift OAuth discovery)
* Added `caFile` to `auth` stanza as a separate CA bundle file for the OAuth endpoints
* Updated to use k8s service (in-cluster) or master public URL (off-cluster) for OpenShift OAuth discovery

coreos-ui commented:
The PR image has been pushed to Quay with tag `pr_65_build_400`

spadgett commented:
Tested in-cluster and off-cluster on both Tectonic and OpenShift clusters.

enj commented on cmd/bridge/config.go:
```diff
@@ -0,0 +1,179 @@
+type Auth struct {
+	ClientID         string `yaml:"clientID"`
+	ClientSecretFile string `yaml:"clientSecretFile"`
+	CAFile           string `yaml:"caFile"`
```
I thought we wanted something like `oauthEndpointCAFile` to be specific.

spadgett replied:
I don't mind the more specific name if you prefer.

spadgett commented:
jenkins push

coreos-ui commented:
The PR image has been pushed to Quay with tag `pr_65_build_401`

spadgett commented:
Can I get an LGTM or further review comments? This is the last thing blocking the install changes from going in, and we're days from feature freeze.

spadgett commented:
jenkins rebuild

ericavonb commented:
@alecmerdler do you have time to give this a quick look over? are there any recent console changes that might affect this?
